module/component/component.go

```diff
@@ -0,0 +1,309 @@
+type OnError = func(err error, triggerRestart func())
+
+func RunComponent(ctx context.Context, componentFactory ComponentFactory, handler OnError) error {
+	// reference to per-run signals for the component
+	var component Component
+	var cancel context.CancelFunc
+	var done <-chan struct{}
+	var irrecoverableErr <-chan error
+
+	start := func() (err error) {
+		component, err = componentFactory()
+		if err != nil {
+			return // failure to generate the component, should be handled out-of-band because a restart won't help
+		}
+
+		// context used to run the component
+		var runCtx context.Context
+		runCtx, cancel = context.WithCancel(ctx)
+
+		// signaler used for irrecoverables
+		signaler := irrecoverable.NewSignaler()
+		signalingCtx := irrecoverable.WithSignaler(runCtx, signaler)
+		irrecoverableErr = signaler.Error()
+
+		if err = component.Start(signalingCtx); err != nil {
+			cancel()
+			return
+		}
+
+		done = component.Done()
+		return
+	}
+
+	shutdownAndWaitForRestart := func(err error) error {
+		// shutdown the component
+		cancel()
+
+		// wait until it's done
+		// note that irrecoverables which are encountered during shutdown are ignored
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-done:
+		}
+
+		// send error to the handler programmed with a restart continuation
+		restartChan := make(chan struct{})
+		go handler(err, func() {
+			close(restartChan)
+		})
+
+		// wait for handler to trigger restart or abort
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-restartChan:
+		}
+
+		return nil
+	}
+
+	for {
+		if err := start(); err != nil {
+			return err // failure to start
+		}
+
+		defer cancel()
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case err := <-irrecoverableErr:
+			if canceled := shutdownAndWaitForRestart(err); canceled != nil {
+				return canceled
+			}
+		}
+	}
+}
+
+// ComponentWorker represents a worker routine of a component
+type ComponentWorker func(ctx irrecoverable.SignalerContext)
+
+// ComponentStartup implements a startup routine for a component
+// This is where a component should perform any necessary startup
+// tasks before worker routines are launched.
+type ComponentStartup func(context.Context) error
+
+// ComponentManagerBuilder provides a mechanism for building a ComponentManager
+type ComponentManagerBuilder interface {
+	// OnStart sets the startup routine for the ComponentManager. If an error is
+	// encountered during the startup routine, the ComponentManager will shutdown
+	// immediately.
+	OnStart(ComponentStartup) ComponentManagerBuilder
+
+	// AddWorker adds a worker routine for the ComponentManager
+	AddWorker(ComponentWorker) ComponentManagerBuilder
+
+	// AddComponent adds a new sub-component for the ComponentManager.
+	// This should be used for critical sub-components whose failure would be
+	// considered irrecoverable. For non-critical sub-components, consider using
+	// RunComponent instead.
```

Comment:
I don't think I get the distinction here. Do you just mean that using `RunComponent` gives you more control over how to handle failures by accepting an error handler, instead of forcing you to implement that explicitly?

Reply:
A component that is suitable to be passed to `RunComponent` is one that is not required to be running for the parent component to function (and hence is okay to restart). If a critical subcomponent fails, the parent component itself is unable to function and therefore also fails.

I think a good way to think about this is with respect to component lifecycles. For example, a component's `Ready` method should never depend on a non-critical subcomponent. If it does, what happens when that non-critical subcomponent fails? Should we go back to non-ready state?

On the other hand, if a subcomponent is required to be running for the parent to function, then the parent's `Ready` method must depend on the subcomponent being ready.

Reply:
I would think of a critical subcomponent basically as just part of the component itself, except encapsulated in a separate `Component` for code clarity / organization reasons. On the other hand, a non-critical sub-component is like a module that can be added or removed.

Reply:
I'm not opposed to the idea of a critical component: the whole idea of having a standardized signaler in a child is to allow a variety of recovery behaviors in a parent. Child components which failure entails that of the parent should be part of the scenery. 

However, a couple of things I'm wondering about the current approach:
- complexity : ideally, `AddComponent` would work with the "same rails" (`Signaler` context, etc), which minimizes cognitive load. We could handle irrecoverables from a critical sub-component easily, but it seems we're adding the "`AddComponent`" lane because we want those components to be run at startup, whereas RunComponent leaves it to the caller. 

- are there always going to be critical components, to the point we need to handle this distinct startup pattern, or is this for the `Middleware`?

I'm not saying there's a clear way to do it, but if we manage to bring those critical subcomponents into the fold of parental supervision, there's a potential to make the workflow of Components easier to use.

Reply:
I see many examples across the codebase where one component's `Ready` method is simply waiting for another component's `Ready`. 

I think it's possible we could simplify the API for component manager by just having a single `AddWorker` API, which takes in a function it can use to signal that it is ready. So the API would look something like this:

```golang
type ComponentWorker func(irrecoverable.SignalerContext, func())

type ComponentManagerBuilder interface {
    AddWorker(ComponentWorker)
}
```

And then the worker routine can contain any logic we want, including calling `RunComponent` to run a component with an error handler. Eg:

```golang
func(ctx irrecoverable.SignalerContext, ready func()) {
    // run some startup logic
    ready()
    
    // do work here, and shutdown when ctx is canceled
    // "work" could include running a subcomponent via RunComponent
}
```

@huitseeker thoughts?

Reply:
Yep, and it matches the continuation-passing approach you could get in `RunComponent` variants. I'm all for it.